mqtt/mqtt-broker/src/error.rs

```diff
@@ -1,135 +1,88 @@
-use std::path;
+use std::path::PathBuf;
 
-use derive_more::Display;
-use failure::{Backtrace, Context, Fail};
 use mqtt3::proto::Packet;
+use thiserror::Error;
 
-#[derive(Debug, Display)]
-pub struct Error {
-    inner: Context<ErrorKind>,
-}
+use crate::Message;
 
-#[derive(Debug, Fail, PartialEq)]
-pub enum ErrorKind {
-    #[fail(display = "An error occurred sending a message to the broker.")]
-    SendBrokerMessage,
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("An error occurred sending a message to the broker.")]
+    SendBrokerMessage(#[source] tokio::sync::mpsc::error::SendError<Message>),
 
-    #[fail(display = "An error occurred sending a message to a connection.")]
-    SendConnectionMessage,
+    #[error("An error occurred sending a message to a connection.")]
+    SendConnectionMessage(#[source] tokio::sync::mpsc::error::SendError<Message>),
 
-    #[fail(display = "An error occurred sending a message to a snapshotter.")]
+    #[error("An error occurred sending a message to a snapshotter.")]
     SendSnapshotMessage,
 
-    #[fail(display = "An error occurred decoding a packet.")]
-    DecodePacket,
+    #[error("An error occurred decoding a packet.")]
+    DecodePacket(#[from] mqtt3::proto::DecodeError),
```

Comment:
Does `#[from]` also set the `#[source]`?

Reply:
yes, but not vice versa. `#[source]` dont generate `impl From`

Reply:
> The #[from] attribute always implies that the same field is #[source], so you don't ever need to specify both attributes.